les/serverpool.go

```diff
@@ -17,904 +17,283 @@
-				}
-				entry.lastConnected = addr
-				entry.addr = make(map[string]*poolEntryAddress)
-				entry.addr[addr.strKey()] = addr
-				entry.addrSelect = *utils.NewWeightedRandomSelect()
-				entry.addrSelect.Update(addr)
-				req.result <- entry
-			}
-
-		case req := <-pool.registerCh:
-			if req.entry.trusted {
-				continue
-			}
-			// Handle peer registration requests.
-			entry := req.entry
-			entry.state = psRegistered
-			entry.regTime = mclock.Now()
-			if !entry.known {
-				pool.newQueue.remove(entry)
-				entry.known = true
-			}
-			pool.knownQueue.setLatest(entry)
-			entry.shortRetry = shortRetryCnt
-			close(req.done)
-
-		case req := <-pool.disconnCh:
-			if req.entry.trusted {
-				continue
-			}
-			// Handle peer disconnection requests.
-			disconnect(req, req.stopped)
-
-		case <-pool.closeCh:
-			if pool.discSetPeriod != nil {
-				close(pool.discSetPeriod)
-			}
-
-			// Spawn a goroutine to close the disconnCh after all connections are disconnected.
-			go func() {
-				pool.connWg.Wait()
-				close(pool.disconnCh)
-			}()
-
-			// Handle all remaining disconnection requests before exit.
-			for req := range pool.disconnCh {
-				disconnect(req, true)
+				return nil, errInvalidField
 			}
-			pool.saveNodes()
-			return
-		}
-	}
-}
+		},
+		func(enc []byte) (interface{}, error) {
+			n := &nodeHistory{}
+			err := rlp.DecodeBytes(enc, &n.dialCost)
+			return n, err
+		},
+	)
+)
 
-func (pool *serverPool) findOrNewNode(node *enode.Node) *poolEntry {
-	now := mclock.Now()
-	entry := pool.entries[node.ID()]
-	if entry == nil {
-		log.Debug("Discovered new entry", "id", node.ID())
-		entry = &poolEntry{
-			node:       node,
-			addr:       make(map[string]*poolEntryAddress),
-			addrSelect: *utils.NewWeightedRandomSelect(),
-			shortRetry: shortRetryCnt,
-		}
-		pool.entries[node.ID()] = entry
-		// initialize previously unknown peers with good statistics to give a chance to prove themselves
-		entry.connectStats.add(1, initStatsWeight)
-		entry.delayStats.add(0, initStatsWeight)
-		entry.responseStats.add(0, initStatsWeight)
-		entry.timeoutStats.add(0, initStatsWeight)
-	}
-	entry.lastDiscovered = now
-	addr := &poolEntryAddress{ip: node.IP(), port: uint16(node.TCP())}
-	if a, ok := entry.addr[addr.strKey()]; ok {
-		addr = a
+// newServerPool creates a new server pool
+func newServerPool(ns *utils.NodeStateMachine, vt *lpc.ValueTracker, discovery enode.Iterator, clock mclock.Clock, trustedURLs []string, testing bool) *serverPool {
+	s := &serverPool{
+		clock: clock,
+		ns:    ns,
+		vt:    vt,
+	}
+	s.getTimeout()
+	// Register all serverpool-defined states
+	stDiscovered := s.ns.MustRegisterState(sfDiscovered)
+	s.stHasValue = s.ns.MustRegisterState(sfHasValue)
+	s.stDialed = s.ns.MustRegisterState(sfDialed)
+	s.stConnected = s.ns.MustRegisterState(sfConnected)
+	s.stRedialWait = s.ns.MustRegisterState(sfRedialWait)
+	s.stAlwaysConnect = s.ns.MustRegisterState(sfAlwaysConnect)
+	s.ns.MustRegisterState(sfSelected)
```

Comment:
Can we also save the `selected flag indicator` here? Then we can pass it to different iterators and they don't need to obtain the indicator anymore